conda/repodata.py

```diff
@@ -0,0 +1,99 @@
+# (c) Continuum Analytics, Inc. / http://continuum.io
+# All Rights Reserved
+#
+# conda is distributed under the terms of the BSD 3-clause license.
+# Consult LICENSE.txt or http://opensource.org/licenses/BSD-3-Clause.
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+from .core import repodata as _impl
+from .connection import CondaSession
+from os.path import join
+
+
+class RepoData(object):
+    """This object represents all the package metainfo of a single channel."""
+
+    _data = {}
+
+    @staticmethod
+    def enable(url, name, priority, cache_dir=None):
+        RepoData._data[url] = RepoData(url, name, priority, cache_dir)
+
+    @staticmethod
+    def get(url):
+        return RepoData._data[url]
+
+    @staticmethod
+    def clean():
+        RepoData._data = {}
+
+    @staticmethod
+    def sync_all(use_cache=False):
+        try:
+            import concurrent.futures
+            with concurrent.futures.ThreadPoolExecutor(10) as e:
+                for rd in RepoData._data.values():
+                    e.submit(rd.sync(use_cache=use_cache, session=CondaSession()))
+        except (ImportError) as e:
+            for rd in RepoData._data.values():
+                rd.sync(use_cache=use_cache, session=CondaSession())
+
+    def __init__(self, url, name, priority, cache_dir=None):
+        """Create a RepoData object."""
+
+        self.url = url
+        self.name = name
+        self.priority = priority
+        self.cache_dir = cache_dir
+        self._data = None
+
+    def sync(self, use_cache=False, session=None):
+        """Syncs this object with an upstream RepoData object."""
+
+        session = session if session else CondaSession()
+        self._data = _impl.fetch_repodata(self.url, self.name, self.priority,
+                                          cache_dir=self.cache_dir,
+                                          use_cache=use_cache, session=session)
+
+    def restore(self, cache_dir=None):
+        """Load data from local cache."""
+
+        cache_path = join(cache_dir or self.cache_dir or _impl.create_cache_dir(),
+                          _impl.cache_fn_url(self.url))
+        mod_etag_headers = _impl.read_mod_and_etag(cache_path)
+        self._data = _impl.read_pickled_repodata(cache_path,
+                                                 self.url, self.name, self.priority,
+                                                 mod_etag_headers.get('_etag'),
+                                                 mod_etag_headers.get('_mod'))
+
+    def persist(self, cache_dir=None):
+        """Save data to local cache."""
+
+        cache_path = join(cache_dir or self.cache_dir or _impl.create_cache_dir(),
+                          _impl.cache_fn_url(self.url))
+        _impl.write_pickled_repodata(cache_path, self._data)
```

Comment:
I'd probably put an `_` in front of `persist()` right now too, to be consistent with what I just said above.